tools/flight_recorder/fr_trace.py

```diff
@@ -626,26 +729,81 @@ def build_collectives(
                         )
                     )
                 nccl_calls.extend(reversed(reversed_calls))
-
         else:
+            has_undecided_case = False
+            errors = Set()
             for o in expected_ranks.intersection(set(other_ranks)):
                 for i, e in enumerate(all_entries[o]):  # type: ignore[index]
                     # step over ops from other PGs
-                    if e["process_group"] == (pg_name, desc):
+                    # only check match state when seq_id matches
+                    if (
+                        e["process_group"] == (pg_name, desc)
+                        and e["collective_seq_id"] == collective_seq_id
+                    ):
+                        match_state = match_one_event(entries[0], e, _memberships)
                         if (
-                            match_one_event(entries[0], e, _memberships)
+                            match_state
+                            in [MatchState.FULLY_MATCHED, MatchState.UNDECIDED]
                             and mismatch[pg_name] == 0
                         ):
                             found_ranks.add(o)
                             found_idx[o] = i
+                            has_undecided_case = match_state == MatchState.UNDECIDED
                         else:
-                            # we found a mismatch. what do we do with that?
-                            mismatch[pg_name] += 1
-                            print(
-                                f"Mismatched collective on rank {o} for group {pg_name}:{desc} collective {profiling_name}"
-                            )
+                            candidate_ranks.add(o)
+                            candidate_idx[o] = i
+                            errors.add(match_state)
                         break
 
+            # case one: not every rank join the collective or in the flight recorder.
+            if (candidate_ranks | found_ranks) != expected_ranks:
+                mismatch[pg_name] += 1
+                print(
+                    f"Not all ranks joining collective for group {pg_name}:{desc} collective {profiling_name}",
+                    f"Missing ranks are {expected_ranks - (candidate_ranks | found_ranks)}",
+                )
+            elif len(candidate_ranks) == 1:
+                # case two: alltoall or alltoall_base case.
```

Comment:
for my education, why was alltoall a tricky case (or why do we mark it UNDECIDED) etc?

Reply:
Because you need to have wait all ranks to check input size and out size match. The input list and output list of alltoall only require the total numel of input and output to be the same.